### PR TITLE
[WIP] MS-DOS port

### PR DESCRIPTION
**Pull request by asiekierka**
Taken https://github.com/MrAlert/megazeux/tree/reDOS-djgpp, removed some hacks and ran with it. 

- [x] Add SB16 audio support.
- [x] Add text page-flipping for EGA renderer.
- [ ] Add character set page-flipping for EGA renderer. (Mildly a stretch goal)
- [ ] Add pre-SB16 audio support. (Stretch goal)
- [ ] Add PC Speaker-only audio support. (Stretch goal)
- [x] Add SVGA bitmapped renderer. (Stretch goal)
- [ ] Add double-buffering for SVGA renderer. (Stretch goal)

- [ ] On slow machines, the audio callback code can run too slowly and thus desynchronize with the SB16 interrupt, causing audible glitches until such a desync happens again. I'd like this not to happen. (Stretch goal; should not happen on real SB16s, however... just clones D:)

- [ ] chdir()ing to "c:/" seems to fail (but doesn't seem to affect anything)

### CODE REVIEW

File: arch/djgpp/audio.c [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+#include <bios.h>
+#include <dpmi.h>
+#include <go32.h>
+#undef delay
+#include "../../src/audio/audio.h"
+#include "platform_djgpp.h"
+
+#ifdef CONFIG_AUDIO
+
+struct sb_config {
```
Comment (AliceLR) [inline arch/djgpp/audio.c]: style and indentation

File: arch/djgpp/audio.c [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+	Uint8 buffer_block;
+	Uint8 old_21h;
+	_go32_dpmi_seginfo old_irq_handler;
+};
+
+static struct sb_config sb_cfg;
+
+static void audio_sb_fill_block(void)
+{
+    int block_size_bytes = audio.buffer_samples << 2;
```
Comment (AliceLR) [inline arch/djgpp/audio.c]: indentation

File: arch/djgpp/audio.c [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    int offset = (sb_cfg.buffer_block != 0) ? block_size_bytes : 0;
+
+    audio_callback((Sint16*) (sb_cfg.buffer + offset), block_size_bytes);
+    dosmemput(sb_cfg.buffer + offset, block_size_bytes, (sb_cfg.buffer_segment << 4) + offset);
+}
+
+static void audio_sb_next_block(void)
+{
+    audio_sb_fill_block();
```
Comment (AliceLR) [inline arch/djgpp/audio.c]: indentation

File: arch/djgpp/audio.c [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+    sb_cfg.buffer_block ^= 1;
+}
+
+static void audio_sb_interrupt(void)
+{
+    audio_sb_next_block();
```
Comment (AliceLR) [inline arch/djgpp/audio.c]: indentation

File: arch/djgpp/audio.c [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+    inportb(sb_cfg.port + 0xF); // ack (sb)
+    outportb(0x20, 0x20); // ack (pic)
+}
+
+static void audio_sb_parse_env(struct sb_config *conf, char *env)
+{
+  char *token;
+  token = strtok(env, " ");
+  while (token != NULL)
```
Comment (AliceLR) [inline arch/djgpp/audio.c]: style

File: arch/djgpp/audio.c [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+  {
+    switch (token[0]) {
```
Comment (AliceLR) [inline arch/djgpp/audio.c]: style

File: arch/djgpp/audio.c [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+        conf->type = strtol(token + 1, NULL, 16);
+        break;
+    }
+    token = strtok(NULL, " ");
+  }
+}
+
+static Uint8 audio_sb_dsp_read(void)
+{
+  while (!(inportb(sb_cfg.port + 0xE) & 0x80));
```
Comment (AliceLR) [inline arch/djgpp/audio.c]: style

File: arch/djgpp/audio.c [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+  return inportb(sb_cfg.port + 0xA);
+}
+
+static void audio_sb_dsp_write(Uint8 val)
+{
+  while (inportb(sb_cfg.port + 0xC) & 0x80);
```
Comment (AliceLR) [inline arch/djgpp/audio.c]: style

File: arch/djgpp/audio.c [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+  outportb(sb_cfg.port + 0xC, val);
+}
+
+static boolean audio_sb_dsp_detect(void)
+{
+  Uint16 i;
+  outportb(sb_cfg.port + 0x6, 1);
+  usleep(3);
+  outportb(sb_cfg.port + 0x6, 0);
+  for (i = 65535; i > 0; i--)
```
Comment (AliceLR) [inline arch/djgpp/audio.c]: style. not commenting the rest in this function.

File: arch/djgpp/audio.c [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+    if (inportb(sb_cfg.port + 0xA) == 0xAA) break;
+  return i > 0;
+}
+
+void init_audio_platform(struct config_info *conf)
+{
+  _go32_dpmi_seginfo new_irq_handler;
+  // Try to find a Sound Blaster
+  char *sb_env = getenv("BLASTER");
+  if (sb_env != NULL)
```
Comment (AliceLR) [inline arch/djgpp/audio.c]: style. not commenting the rest in this function or I'd be doing this all day.

File: arch/djgpp/audio.c [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+      audio_sb_dsp_write(((audio.buffer_samples << 1) - 1) & 0xFF);
+      audio_sb_dsp_write(((audio.buffer_samples << 1) - 1) >> 8);
+    }
+  }
+}
+
+void quit_audio_platform(void)
+{
+  // Deinitialize audio
+  if (sb_cfg.port != 0)
```
Comment (AliceLR) [inline arch/djgpp/audio.c]: style

File: arch/djgpp/mutex.h [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+static inline boolean platform_mutex_lock(platform_mutex *mutex)
+{
+  *mutex = disable();
+  return true;
+}
+
+static inline boolean platform_mutex_unlock(platform_mutex *mutex)
+{
+  if (*mutex)
```
Comment (AliceLR) [inline arch/djgpp/mutex.h]: style

File: arch/djgpp/platform.c [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+int djgpp_display_adapter_detect(void)
+{
+  __dpmi_regs reg;
+
+  // VESA SuperVGA BIOS (VBE) - GET SuperVGA INFORMATION
+  // Generally supported by SVGA cards
+  reg.x.ax = 0x4F00;
+  __dpmi_int(0x10, &reg);
+
+  if (reg.x.ax == 0x004F)
```
Comment (AliceLR) [inline arch/djgpp/platform.c]: style

File: arch/djgpp/platform.c [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+    return DISPLAY_ADAPTER_SVGA;
+
+  // VIDEO - GET DISPLAY COMBINATION CODE (PS,VGA/MCGA)
+  // Generally supported by VGA cards
+  reg.x.ax = 0x1A00;
+  __dpmi_int(0x10, &reg);
+
+  if (reg.h.al == 0x1A)
```
Comment (AliceLR) [inline arch/djgpp/platform.c]: style

File: arch/djgpp/platform.c [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+  {
+    switch (reg.h.bl) {
```
Comment (AliceLR) [inline arch/djgpp/platform.c]: style

File: arch/djgpp/platform.c [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+    }
+  }
+
+  // VIDEO - ALTERNATE FUNCTION SELECT (PS, EGA, VGA, MCGA) - GET EGA INFO
+  // Generally supported by EGA cards
+  reg.h.ah = 0x12;
+  reg.x.bx = 0xFF10;
+  __dpmi_int(0x10, &reg);
+
+  if (reg.h.bh != 0xFF)
```
Comment (AliceLR) [inline arch/djgpp/platform.c]: style

File: arch/djgpp/platform.c [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+{
+  return disp_adapter_names[adapter];
+}
+
+int djgpp_malloc_boundary(int len_bytes, int boundary_bytes, int *selector)
+{
+  int len_segment = (len_bytes + 15) >> 4;
+  int boundary_mask = ~((boundary_bytes - 1) >> 4);
+  int segment = __dpmi_allocate_dos_memory((len_bytes + 7) >> 3, selector);
+  if (((segment + len_segment - 1) & boundary_mask) != (segment & boundary_mask))
```
Comment (AliceLR) [inline arch/djgpp/platform.c]: style

File: arch/djgpp/render_ega.c [hunk 18]
```diff
@@ -0,0 +1,438 @@
...
+  outportb(0x03C5, 0x02);
+  outportb(0x03C4, 0x02);
+  outportb(0x03C5, 0x03);
+  outportb(0x03CE, 0x04);
+  outportb(0x03CF, 0x00);
+}
+
+static void ega_vsync(void)
+{
+  while (inportb(0x03DA) & 0x08);
```
Comment (AliceLR) [inline arch/djgpp/render_ega.c]: style

File: arch/djgpp/event.c [hunk 19]
```diff
@@ -0,0 +1,583 @@
...
+    {
+      input.key_repeat_stack[input.repeat_stack_pointer] =
+       status->key_repeat;
+      input.unicode_repeat_stack[input.repeat_stack_pointer] =
+       status->unicode_repeat;
+      input.repeat_stack_pointer++;
+    }
+  }
+
+  key_press(status, ikey, unicode);
```
Comment (AliceLR) [inline arch/djgpp/event.c:369]: Whoops, missed one. This needs to be changed to
```
  key_press(status, ikey);
  key_press_unicode(status, unicode, true);
```
to build with the lastest master.

**Comment by AliceLR** [comment]
Other minor things:

To correct the path slashes, change line 32 of `src/io/path.h` to 
```C
#if defined(__WIN32__) || defined(CONFIG_DJGPP)
```

To fix the missing drive letters, add this at line 3501 of `src/window.c`:
```C
#ifdef CONFIG_DJGPP
    if(dirs_okay == 1)
    {
      int current_disk = getdisk();
      int max_disk = setdisk(current_disk);
      for(i = 0; i < max_disk; i++)
      {
        setdisk(i);
        if(getdisk() != i)
          continue;

        dir_list[num_dirs] = cmalloc(3);
        sprintf(dir_list[num_dirs], "%c:", 'A' + i);
        num_dirs++;

        if(num_dirs == total_dirnames_allocated)
        {
          dir_list = crealloc(dir_list, sizeof(char *) *
           total_dirnames_allocated * 2);
          memset(dir_list + total_dirnames_allocated, 0,
           sizeof(char *) * total_dirnames_allocated);
          total_dirnames_allocated *= 2;
        }
      }
      setdisk(current_disk);
    }
#endif
```

**Comment by AliceLR** [comment]
Here's a patch that provides basic support for truncating LFNs in newer games to DOS SFNs. There are about a zillion issues with doing this but the alternative is games like [Arc](https://www.digitalmzx.com/show.php?id=1748) just don't play any music at all. Not really advocating for or against this being included right now, but I suspect it will fix many games relying on LFNs.

Note that another alternative if MZX is just being run in DOSBox is to use something like DOSBox-X instead (which has LFN support).

```diff
diff --git a/src/io/fsafeopen.c b/src/io/fsafeopen.c
index f0a8e59f..06e6e195 100644
--- a/src/io/fsafeopen.c
+++ b/src/io/fsafeopen.c
@@ -311,18 +311,25 @@ static int case5(char *path, size_t buffer_len, char *string, boolean check_sfn)
     const char *string_cmp = string;
     char string_sfn[SFN_BUFFER_LEN];
     char newpath_sfn[SFN_BUFFER_LEN];
-    boolean string_is_wildcard_sfn = false;
     boolean has_sfn_match = false;
+    enum sfn_type type;

     // If the input path is a truncated SFN, it may need to be aggressively
     // checked against generated SFNs from files in the directory.
-    if(check_sfn && is_sfn(string, strlen(string)) == SFN_TRUNCATED)
+    type = is_sfn(string, strlen(string));
+    if(check_sfn && type == SFN_TRUNCATED)
     {
       memcpy(string_sfn, string, strlen(string) + 1);
-      string_is_wildcard_sfn = true;
       string_cmp = string_sfn;
     }

+#ifdef CONFIG_DJGPP
+    if(type == NOT_AN_SFN)
+    {
+      get_sfn(string_sfn, SFN_BUFFER_LEN, string);
+    }
+#endif
+
     while(dir_get_next_entry(&wd, newpath, NULL))
     {
       // okay, we got something, but does it match?
@@ -334,7 +341,23 @@ static int case5(char *path, size_t buffer_len, char *string, boolean check_sfn)
       }
       else

-      if(string_is_wildcard_sfn)
+#ifdef CONFIG_DJGPP
+      // Is there a file matching the generated SFN for the requested LFN?
+      if(type == NOT_AN_SFN)
+      {
+        if(!strcasecmp(string_sfn, newpath))
+        {
+          trace("%s:%d: truncated LFN '%s' to '%s'\n",
+           __FILE__, __LINE__, string, newpath);
+          memcpy(string, newpath, strlen(newpath) + 1);
+          ret = FSAFE_SUCCESS;
+          break;
+        }
+      }
+      else
+#endif
+
+      if(type == SFN_TRUNCATED)
       {
         const char *newpath_cmp = get_sfn(newpath_sfn, SFN_BUFFER_LEN, newpath);
         if(!strcasecmp(string_cmp, newpath_cmp))
```

**Comment by AliceLR** [comment]
Patch to fix building due to recent changes in master. I don't know if there's really an appropriate way of implementing these but I doubt it matters much anyway, so I just copied the `thread_dummy.h` implementations.

```diff
diff --git a/arch/djgpp/mutex.h b/arch/djgpp/mutex.h
index bf8b1b2a..0d17f2f7 100644
--- a/arch/djgpp/mutex.h
+++ b/arch/djgpp/mutex.h
@@ -31,6 +31,7 @@
 __M_BEGIN_DECLS
 
 typedef int platform_mutex;
+typedef int platform_thread_id;
 
 static inline void platform_mutex_init(platform_mutex *mutex)
 {
@@ -53,6 +54,17 @@ static inline boolean platform_mutex_unlock(platform_mutex *mutex)
   return true;
 }
 
+static inline platform_thread_id platform_get_thread_id(void)
+{
+  return 0;
+}
+
+static inline boolean platform_is_same_thread(platform_thread_id a,
+ platform_thread_id b)
+{
+  return true;
+}
+
 __M_END_DECLS
 
 #endif // __MUTEX_DJGPP_H
```

**Comment by AliceLR** [comment]
Patch to fix a warning with the latest master caused by `arch/djgpp/audio.c` using the worst string function in the entire standard library.
```diff
diff --git a/arch/djgpp/audio.c b/arch/djgpp/audio.c
index 0cae810d..f88cf7ee 100644
--- a/arch/djgpp/audio.c
+++ b/arch/djgpp/audio.c
@@ -80,9 +80,9 @@ static void audio_sb_interrupt(void)
 static void audio_sb_parse_env(struct sb_config *conf, char *env)
 {
   char *token;
-  token = strtok(env, " ");
-  while(token != NULL)
+  while(env != NULL)
   {
+    token = strsep(&env, " ");
     switch(token[0])
     {
       case 'A':
@@ -101,7 +101,6 @@ static void audio_sb_parse_env(struct sb_config *conf, char *env)
         conf->type = strtol(token + 1, NULL, 16);
         break;
     }
-    token = strtok(NULL, " ");
   }
 }

```

File: arch/djgpp/audio.c [hunk 20]
```diff
@@ -0,0 +1,281 @@
...
+    if(!sb_cfg.nearptr_buffer_enabled)
+    {
+      sb_cfg.buffer = malloc(buffer_size_bytes);
+      memset(sb_cfg.buffer, 0, buffer_size_bytes);
+      dosmemput(sb_cfg.buffer, buffer_size_bytes, sb_cfg.buffer_segment << 4);
+    }
+
+    // lock C irq handler
+    // (TODO: rewrite handler in ASM?)
+    _go32_dpmi_lock_code(audio_sb_interrupt, 1024);
```
Comment (AliceLR) [inline arch/djgpp/audio.c:225]: It'd be nice to make this warning die but I'm not sure what short of adding a pragma to disable `-Wpedantic`.
```
  CC       arch/djgpp/audio.c
arch/djgpp/audio.c: In function 'init_audio_platform':
arch/djgpp/audio.c:224:26: warning: ISO C forbids passing argument 1 of '_go32_dpmi_lock_code' between function pointer and 'void *' [-Wpedantic]
  224 |     _go32_dpmi_lock_code(audio_sb_interrupt, 1024);
      |                          ^~~~~~~~~~~~~~~~~~
In file included from arch/djgpp/audio.c:26:
d:\djgpp\i586-pc-msdosdjgpp\sys-include\dpmi.h:337:33: note: expected 'void *' but argument is of type 'void (*)(void)'
  337 | int _go32_dpmi_lock_code( void *_lockaddr, unsigned long _locksize);
      |                           ~~~~~~^~~~~~~~~
```

**Comment by AliceLR** [comment]
Here's a patch for the latest master that fixes all of the cursor display bugs as far as I can tell.
```diff
diff --git a/arch/djgpp/render_ega.c b/arch/djgpp/render_ega.c
index cca32e57..3d8edc87 100644
--- a/arch/djgpp/render_ega.c
+++ b/arch/djgpp/render_ega.c
@@ -412,8 +412,8 @@ static void ega_render_graph(struct graphics_data *graphics)
                  | (src->char_value & 0xFF));
 }
 
-static void ega_render_cursor(struct graphics_data *graphics,
- Uint32 x, Uint32 y, Uint16 color, Uint8 lines, Uint8 offset)
+static void ega_hardware_cursor(struct graphics_data *graphics,
+ Uint32 x, Uint32 y, Uint16 color, Uint8 lines, Uint8 offset, boolean enable)
 {
   struct ega_render_data *render_data = graphics->render_data;
   unsigned long dest = ega_vb_page[render_data->page];
@@ -426,7 +426,7 @@ static void ega_render_cursor(struct graphics_data *graphics,
     render_data->offset = offset;
     render_data->curpages = 4;
   }
-  if(lines)
+  if(enable)
   {
     if((x != render_data->x) || (y != render_data->y))
     {
@@ -439,8 +439,14 @@ static void ega_render_cursor(struct graphics_data *graphics,
       ega_set_cursor_pos(render_data->page, x, y);
       render_data->curpages--;
     }
-    _farsetsel(render_data->vbsel);
-    _farnspokeb(dest, (_farnspeekb(dest) & 0xF0) | (color & 0x0F));
+    /**
+     * Disable setting the cursor color for now. This changes the color of the
+     * color of the entire char which is generally not desirable; the point of
+     * having a different cursor color than the char color is to make it more
+     * visible.
+     */
+    //_farsetsel(render_data->vbsel);
+    //_farnspokeb(dest, (_farnspeekb(dest) & 0xF0) | (color & 0x0F));
   }
 }
 
@@ -489,7 +495,7 @@ void render_ega_register(struct renderer *renderer)
   renderer->get_screen_coords = get_screen_coords_centered;
   renderer->set_screen_coords = set_screen_coords_centered;
   renderer->render_graph = ega_render_graph;
-  renderer->render_cursor = ega_render_cursor;
+  renderer->hardware_cursor = ega_hardware_cursor;
   renderer->render_mouse = ega_render_mouse;
   renderer->sync_screen = ega_sync_screen;
 }
```

File: arch/djgpp/audio.c (hunk 20 above)
Comment (asiekierka) [reply arch/djgpp/audio.c:224]: I would just leave it alone, the specific thing it does is a *very* ugly hack anyway...

Comment (AliceLR) [reply arch/djgpp/audio.c:224]: Looking [more into go32](http://www.delorie.com/djgpp/v2faq/faq18_10.html), yeah, I agree that this warning should stick around for now.

**Comment by AliceLR** [comment]
Here's a patch that makes the DJGPP port attempt chdir to the startup dir on exit for convenience.
```
diff --git a/src/main.c b/src/main.c
index 1238e10a..0a372eb7 100644
--- a/src/main.c
+++ b/src/main.c
@@ -365,5 +365,8 @@ err_free_res:
   mzx_res_free();
   platform_quit();
 err_out:
+#ifdef CONFIG_DJGPP
+  chdir(startup_dir);
+#endif
   return err;
 }
```

Patch to set both resampling modes to nearest by default. This doesn't fix the audio problems but it should make them noticeably less bad (at the cost of the audio sounding like... well, DOS game audio).
```
diff --git a/src/configure.c b/src/configure.c
index 3ff427aa..5f2ac467 100644
--- a/src/configure.c
+++ b/src/configure.c
@@ -95,6 +95,11 @@
 #define AUTO_DECRYPT_WORLDS true
 #endif
 
+#ifdef CONFIG_DJGPP
+#define RESAMPLE_MODE_DEFAULT RESAMPLE_MODE_NONE
+#define MOD_RESAMPLE_MODE_DEFAULT RESAMPLE_MODE_NONE
+#endif
+
 // End arch-specific config.
 
 #ifndef FORCE_BPP_DEFAULT
@@ -123,6 +128,14 @@
 #define AUDIO_SAMPLE_RATE 44100
 #endif
 
+#ifndef RESAMPLE_MODE_DEFAULT
+#define RESAMPLE_MODE_DEFAULT RESAMPLE_MODE_LINEAR
+#endif
+
+#ifndef MOD_RESAMPLE_MODE_DEFAULT
+#define MOD_RESAMPLE_MODE_DEFAULT RESAMPLE_MODE_CUBIC
+#endif
+
 #ifndef FULLSCREEN_WIDTH_DEFAULT
 #define FULLSCREEN_WIDTH_DEFAULT -1
 #endif
@@ -202,8 +215,8 @@ static const struct config_info user_conf_default =
   AUDIO_SAMPLE_RATE,            // output_frequency
   AUDIO_BUFFER_SAMPLES,         // audio_buffer_samples
   0,                            // oversampling_on
-  RESAMPLE_MODE_LINEAR,         // resample_mode
-  RESAMPLE_MODE_CUBIC,          // module_resample_mode
+  RESAMPLE_MODE_DEFAULT,        // resample_mode
+  MOD_RESAMPLE_MODE_DEFAULT,    // module_resample_mode
   -1,                           // max_simultaneous_samples
   8,                            // music_volume
   8,                            // sam_volume
```

...the audio mutex wasn't getting destroyed until a couple of commits ago. Quick patch so this port will build, though it'd really probably be better eventually to handle `audio_callback` in `delay` and have this port use `thread_dummy.h`.
```
diff --git a/arch/djgpp/mutex.h b/arch/djgpp/mutex.h
index 0d17f2f7..a97cddb0 100644
--- a/arch/djgpp/mutex.h
+++ b/arch/djgpp/mutex.h
@@ -38,6 +38,11 @@ static inline void platform_mutex_init(platform_mutex *mutex)
   *mutex = 0;
 }
 
+static inline void platform_mutex_destroy(platform_mutex *mutex)
+{
+  *mutex = 0;
+}
+
 static inline boolean platform_mutex_lock(platform_mutex *mutex)
 {
   *mutex = disable();
```

**Comment by AliceLR** [comment]
I think I added a merge conflict a few commits ago (probably reindenting the config.sh options table), can you rebase or merge master. Also, I think the DOS SVGA option should probably go in the graphics options section (like the GP2X renderer... I think the sections were added well after the DOS SVGA option already existed in this branch).

**Comment by AliceLR** [comment]
Wow, since the Makefile merge conflict is nontrivial (sorry), here's what that section of the Makefile looks like for me after I resolved it:
```Makefile
#
# The following flags are not applicable to mingw or djgpp builds.
#
ifneq (${PLATFORM},mingw)
ifneq (${PLATFORM},djgpp)

#
# Symbols in COFF binaries are implicitly hidden unless exported; this
# flag just confuses GCC and must be disabled.
#
ifeq (${HAS_F_VISIBILITY},1)
CFLAGS   += -fvisibility=hidden
CXXFLAGS += -fvisibility=hidden
endif

endif # PLATFORM=djgpp
endif # PLATFORM=mingw

#
# The stack protector is optional and is generally only built for Linux/BSD and
# Mac OS X. Windows and most embedded platforms currently disable this.
#
ifeq (${BUILD_STACK_PROTECTOR},1)
ifeq (${HAS_F_STACK_PROTECTOR},1)
CFLAGS   += -fstack-protector-all
CXXFLAGS += -fstack-protector-all
else
$(warning stack protector not supported, ignoring.)
endif
endif

#
# Enable -fanalyzer if supported.
#

```

**Comment by AliceLR** [comment]
Wow, nevermind, I guess GitHub lets me commit to your PR now...

File: src/window.c [hunk 21]
```diff
@@ -3512,6 +3517,34 @@ __editor_maybe_static int file_manager(struct world *mzx_world,
     }
 #endif
 
+#ifdef CONFIG_DJGPP
+    if(dirs_okay == 1)
```
Comment (AliceLR) [inline src/window.c:3521]: I broke this by merging #286; this line needs to be changed to:
```C
    if(allow_dirs == ALLOW_ALL_DIRS)
```

**Comment by flarn2006** [comment]
Wasn't this already on DOS before it was anywhere else?

**Comment by asiekierka** [comment]
Yes, but it was ported away from DOS because by the mid-00s people weren't really running DOS (or Windows 9x) natively anymore.

**Review by AliceLR — APPROVED** [review]

**Pull request merged**